### PR TITLE
[FSDP] Ignore buffer type casting in ignored modules

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106766

issue resolved: https://github.com/pytorch/pytorch/issues/97791

before this PR, mixed_precision applies to buffers from ignored modules. see ```test_state_dict_with_ignored_modules(mixed_precision=True)``` for reproduce

after, we avoid applying mixed_precision semantics to buffers from ignored modules
* step 1 initialization: state._ignored_buffer_names contains all the buffers from ignored modules
* step 2 lazy init at runtime: skip ignored buffers in ```_get_buffers_and_dtypes_for_computation```
* step 3 skip upcasting in state_dict hook: avoid upcasting for ignored buffers in ```_get_buffers_and_dtypes_for_computation```

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 1]
```diff
@@ -1022,6 +1056,13 @@ def test_state_dict_with_ignored_modules(
                 sd1[prefixed_tensor_name].data_ptr(),
                 f"{prefixed_tensor_name}",
             )
+        # should not apply mixed_precision to ignored buffers
+        for buffer_name in {
+            **buffer_to_buffer_name,
+        }.values():
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_state_dict.py:1062]: nit: Can this be simplified to:
```suggestion
        for buffer_name in buffer_to_buffer_name.values():
```

File: torch/distributed/fsdp/_init_utils.py [hunk 2]
```diff
@@ -635,6 +639,37 @@ def _get_ignored_params(
     return all_ignored_params
 
 
+def _get_ignored_buffer_names(
+    root_module: torch.nn.Module,
+    ignored_modules: Set[torch.nn.Module],
+) -> Set[str]:
+    """
+    Returns the buffers of the modules in ``ignored_modules``
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:647]:
```suggestion
    Returns the cleaned buffer FQNs of the modules in ``ignored_modules``
```

File: torch/distributed/fsdp/_init_utils.py [hunk 3]
```diff
@@ -635,6 +639,37 @@ def _get_ignored_params(
...
+
+    all_ignored_buffer_names.update(
+        {
+            clean_tensor_name(buffer_name)
+            for buffer_name, buffer in root_module.named_buffers()
+            if buffer in buffers_in_ignored_modules
+        }
+    )
+
+    # Always include nested FSDP modules' ignored buffers
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:663]: nit:
```suggestion
    # Always include nested FSDP modules' ignored buffer names
```

File: torch/distributed/fsdp/_common_utils.py [hunk 4]
```diff
@@ -111,6 +111,7 @@ def __init__(self) -> None:
         # FSDP/fully_shard.
         self._ignored_modules: Set[nn.Module] = set()
         self._ignored_params: Set[nn.Parameter] = set()
+        self._ignored_buffer_names: Set[str] = set()
```
Comment (awgu) [inline torch/distributed/fsdp/_common_utils.py:114]:
```suggestion
        # Buffer names are cleaned (without wrapper prefixes)
        self._ignored_buffer_names: Set[str] = set()
```

File: torch/distributed/fsdp/_runtime_utils.py [hunk 5]
```diff
@@ -1485,10 +1486,15 @@ def _get_buffers_and_dtypes_for_computation(
         root_module
     )
     for fsdp_state, fsdp_module in zip(reversed(fsdp_states), reversed(fsdp_modules)):
-        for buffer in fsdp_module.buffers():
+        for buffer_name, buffer in fsdp_module.named_buffers():
             if buffer in visited_buffers:
                 continue
             visited_buffers.add(buffer)
+            if (
+                hasattr(fsdp_state, "_ignored_buffer_names")
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:1494]: Do we need the `hasattr(fsdp_state, "_ignored_buffer_names")` check? This attribute should always be defined?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 6]
```diff
@@ -250,8 +250,14 @@ def _common_unshard_post_state_dict_hook(
                 and buffer.device != cpu_device
             ):
                 state_dict[fqn] = buffer.to(cpu_device)
-            buffer_clean_fqns.append(clean_key)
-            buffers.append(state_dict[fqn])
+            # skip upcasting for ignored buffers
+            if (
+                hasattr(fsdp_state, "_ignored_buffer_names")
```
Comment (awgu) [inline torch/distributed/fsdp/_state_dict_utils.py:255]: Same with this `hasattr` check

**Review by awgu — APPROVED** [review]
This looks good to me! Thanks for the fast PR. I left some minor nits. The main thing is that for `FullyShardedDataParallel` (which is an `nn.Module` wrapper), we care about "clean" vs. "unclean" names, so it is good to be clear which kind of names we are using.

"Clean" means that any `nn.Module` wrapper prefixes have been removed (e.g. `_fsdp_wrapped_module.`). "Unclean" means they are left there untouched.

Also, JFYI, FQN stands for fully-qualified name, and this usually refers to the name returned from `named_parameters()` / `named_buffers()` when called on a root module. Sometimes though, it can be ambiguous which module the name is prefixed from, in which case it is good to make that clear in code/comments.

**Comment by weifengpy** [comment]
> 

thanks for the context. FQN explanation really helped. Before this, I only get to understand it after pdb prints.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 5 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_runtime_utils.py]: you are right. we do not need hasattr check. removed the check

**Pull request closed without merging**